README.md
# Farmpy
Python3 package to handle job submission to a compute farm

[![Build Status](https://travis-ci.org/sanger-pathogens/Farmpy.svg?branch=master)](https://travis-ci.org/sanger-pathogens/Farmpy)   
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/sanger-pathogens/Farmpy/blob/master/LICENSE)

## Contents
  * [Introduction](#introduction)
  * [Installation](#installation)
    * [Required dependencies](#required-dependencies)
    * [Using pip](#using-pip)
    * [Run the tests:](#run-the-tests)
  * [Usage \- command line](#usage---command-line)
    * [Submitting jobs](#submitting-jobs)
    * [Getting stats from finished jobs](#getting-stats-from-finished-jobs)
  * [Usage \- running within a script](#usage---running-within-a-script)
    * [Make a job and run it](#make-a-job-and-run-it)
    * [Dependencies](#dependencies)
    * [Job arrays](#job-arrays)
  * [License](#license)
  * [Feedback/Issues](#feedbackissues)

## Introduction
This is a Python3 package to handle job submission to a compute farm. Currently supports Platform's LSF. It has a script to make submitting jobs easier and also an API to run jobs from within scripts.

## Installation
Farmpy has the following dependencies:

### Required dependencies
We assume that you are using Platform's LSF, i.e. it is installed already with bsub and lsadmin in your path.

Details for installing Farmpy are provided below. If you encounter an issue when installing Farmpy please contact your local system administrator. If you encounter a bug please log it [here](https://github.com/sanger-pathogens/Farmpy/issues) or email us at path-help@sanger.ac.uk.

### Using pip

`python3 setup.py install`

### Run the tests:
The test can be run from the top level directory:  

`python3 setup.py test`

## Usage - command line

### Submitting jobs

To submit a job called "name" that asks for 1GB of memory, runs the script foo.sh and writes stdout and stderr to name.o and name.e:

`bsub.py 1 name foo.sh`

To submit a job that will only run when job with ID 42 has finished, use:

`bsub.py --done 42 1 name foo.sh`

If you want 10GB of /tmp space on the node:

`bsub.py --tmp_space 10 1 name foo.sh`

There are many more options. Use -h or --help to see the full list of options

`bsub.py --help`

### Getting stats from finished jobs

To get a tsv file of CPU, memory usage etc of a job:

`bsub_out_to_stats bsub.output`

This works with muliple files. e.g.:

`bsub_out_to_stats bsub.output1 bsub.output2`

or

`bsub_out_to_stats *.output`

Each file can contain the output of more than one job.

## Usage - running within a script

Amongst other things, you run a job, set dependencies, change queues and resources and run arrays. Use

`help(lsf)`

to find out more.

### Make a job and run it

```
job1 = lsf.Job('out', 'err', 'name', 'normal', 1, 'foo.sh')
job1.run()
```

This created a job called 'name', which will run in the normal queue.  It asked for 1GB of memory and will write stdout/stderr to the files out/err.  It runs the script run.sh.

### Dependencies

When the job was run, the variable job1.job_id was set, so that you can use dependencies. For example:
```
job2 = lsf.Job('out2', 'err2', 'job2', 'normal', 1, 'run2.sh')
job2.add_dependency(job1.job_id)
job2.run()
```

Alternatively, dependcies can be set using job names, but these might not be unique so IDs are safer. If you want to use a name instead:

`job2.add_dependency(job1.name)`

### Job arrays

You can run a job array using start= and end= when constructing a Job. Example:

`job = ('out', 'err', 'name', 'normal', 1, 'run.sh INDEX', start=1, end=10)`

This sets up a job array with 10 elements. stdout files will be called out.1, out.2, ....etc and similarly for stderr. Every appearance of 'INDEX' in the command is translated to be the job index (technically, the $LSB_JOBINDEX environment variable). So this would submit 10 jobs to LSF:
```
    run.sh 1
    run.sh 2
    ...
    run.sh 10
```
## License
Farmpy is free software, licensed under [GPLv3](https://github.com/sanger-pathogens/Farmpy/blob/master/LICENSE).

## Feedback/Issues
Please report any issues to the [issues page](https://github.com/sanger-pathogens/Farmpy/issues) or email path-help@sanger.ac.uk.


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/sanger-pathogens/Farmpy/tree/HEAD)

[Full Changelog](https://github.com/sanger-pathogens/Farmpy/compare/v0.42.0...HEAD)

**Merged pull requests:**

- Update readme [\#17](https://github.com/sanger-pathogens/Farmpy/pull/17) ([ssjunnebo](https://github.com/ssjunnebo))

## [v0.42.0](https://github.com/sanger-pathogens/Farmpy/tree/v0.42.0) (2018-10-25)
[Full Changelog](https://github.com/sanger-pathogens/Farmpy/compare/v0.41.0...v0.42.0)

**Merged pull requests:**

- get memory units from environment variable [\#16](https://github.com/sanger-pathogens/Farmpy/pull/16) ([martinghunt](https://github.com/martinghunt))
- 574328 add travis [\#15](https://github.com/sanger-pathogens/Farmpy/pull/15) ([ssjunnebo](https://github.com/ssjunnebo))

## [v0.41.0](https://github.com/sanger-pathogens/Farmpy/tree/v0.41.0) (2017-08-11)
[Full Changelog](https://github.com/sanger-pathogens/Farmpy/compare/v0.40...v0.41.0)

**Merged pull requests:**

- Array limit option [\#14](https://github.com/sanger-pathogens/Farmpy/pull/14) ([martinghunt](https://github.com/martinghunt))

## [v0.40](https://github.com/sanger-pathogens/Farmpy/tree/v0.40) (2017-05-25)
[Full Changelog](https://github.com/sanger-pathogens/Farmpy/compare/v0.3.2...v0.40)

**Merged pull requests:**

- Default no queue [\#13](https://github.com/sanger-pathogens/Farmpy/pull/13) ([martinghunt](https://github.com/martinghunt))
- update AUTHORS and LICENSE [\#12](https://github.com/sanger-pathogens/Farmpy/pull/12) ([ssjunnebo](https://github.com/ssjunnebo))

## [v0.3.2](https://github.com/sanger-pathogens/Farmpy/tree/v0.3.2) (2017-01-09)
[Full Changelog](https://github.com/sanger-pathogens/Farmpy/compare/v0.3.1...v0.3.2)

**Merged pull requests:**

- Remove long description [\#11](https://github.com/sanger-pathogens/Farmpy/pull/11) ([martinghunt](https://github.com/martinghunt))

## [v0.3.1](https://github.com/sanger-pathogens/Farmpy/tree/v0.3.1) (2016-03-29)
**Merged pull requests:**

- Reported at to on [\#10](https://github.com/sanger-pathogens/Farmpy/pull/10) ([martinghunt](https://github.com/martinghunt))
- Add support for tokens [\#9](https://github.com/sanger-pathogens/Farmpy/pull/9) ([martinghunt](https://github.com/martinghunt))
- Added AUTHORS file [\#8](https://github.com/sanger-pathogens/Farmpy/pull/8) ([aslett1](https://github.com/aslett1))
- Add checkpointing [\#7](https://github.com/sanger-pathogens/Farmpy/pull/7) ([martinghunt](https://github.com/martinghunt))
- Report all times in hours, with option to use seconds instead [\#6](https://github.com/sanger-pathogens/Farmpy/pull/6) ([martinghunt](https://github.com/martinghunt))
- Add script to get stats from output files. Bug fix with queue not getting set [\#5](https://github.com/sanger-pathogens/Farmpy/pull/5) ([martinghunt](https://github.com/martinghunt))
- added method to run job directly instead of bsubbing it [\#4](https://github.com/sanger-pathogens/Farmpy/pull/4) ([martinghunt](https://github.com/martinghunt))
- fix filename in long\_description [\#3](https://github.com/sanger-pathogens/Farmpy/pull/3) ([martinghunt](https://github.com/martinghunt))
- fixed typos in the readme [\#2](https://github.com/sanger-pathogens/Farmpy/pull/2) ([martinghunt](https://github.com/martinghunt))
- Initial add of files [\#1](https://github.com/sanger-pathogens/Farmpy/pull/1) ([martinghunt](https://github.com/martinghunt))




farmpy/tests/lsf_test.py
#!/usr/bin/env python3

import sys
import unittest
import filecmp
from farmpy import lsf
import os

lsf_dir = os.path.dirname(os.path.abspath(lsf.__file__))
test_dir = os.path.join(lsf_dir, 'tests', 'data')

class TestJob(unittest.TestCase):
    def test_make_prexec_test_string(self):
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        home = os.path.expanduser('~')
        self.assertEqual("-E 'test -e " + home + "'", bsub._make_prexec_test_string())

    def test_set_memory_units(self):
        '''Check that we can get memory units or die gracefully'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')

        with self.assertRaises(lsf.Error):
            bsub._lsadmin_cmd = 'this_is_not_a_command_and_should_cause_error'
            bsub._set_memory_units()

        with self.assertRaises(lsf.Error):
            bsub._lsadmin_cmd = 'cat ' + os.path.join(test_dir, 'lsf_unittest_lsadmin_showconf_bad_units.txt')
            bsub._set_memory_units()

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._lsadmin_cmd = 'cat ' + os.path.join(test_dir, 'lsf_unittest_lsadmin_showconf_no_units.txt')
        bsub._set_memory_units()
        self.assertEqual('KB', bsub.memory_units)

        os.environ['FARMPY_LSF_MEMORY_UNITS'] = 'MB'
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._lsadmin_cmd = 'cat ' + os.path.join(test_dir, 'lsf_unittest_lsadmin_showconf_kb.txt')
        bsub._set_memory_units()
        self.assertEqual('MB', bsub.memory_units)
        del os.environ['FARMPY_LSF_MEMORY_UNITS']

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._lsadmin_cmd = 'cat ' + os.path.join(test_dir, 'lsf_unittest_lsadmin_showconf_kb.txt')
        bsub._set_memory_units()
        self.assertEqual('KB', bsub.memory_units)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._lsadmin_cmd = 'cat ' + os.path.join(test_dir, 'lsf_unittest_lsadmin_showconf_mb.txt')
        bsub._set_memory_units()
        self.assertEqual('MB', bsub.memory_units)

    def test_make_checkpoint_string(self):
        '''Test make_checkpoint_string'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        self.assertEqual('', bsub._make_checkpoint_string())

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', checkpoint=True)
        self.assertEqual('-k "' + os.path.abspath('out.checkpoint') + ' method=blcr 600"', bsub._make_checkpoint_string())

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', checkpoint=True, checkpoint_dir='foo', checkpoint_period=42)
        self.assertEqual('-k "' + os.path.abspath('foo') + ' method=blcr 42"', bsub._make_checkpoint_string())

    def test_make_resources_string(self):
        '''Check resources in bsub call get set correctly'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1.5, 'cmd', no_resources=True)
        self.assertEqual('', bsub._make_resources_string())

        # need to manually set the lsf units for these tests
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1.5, 'cmd')
        bsub.memory_units='MB'
        resources = bsub._make_resources_string()
        self.assertEqual('-R "select[mem>1500] rusage[mem=1500]" -M1500', resources)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1.5, 'cmd', tokens_name='mytokens', tokens_number=42)
        bsub.memory_units='MB'
        resources = bsub._make_resources_string()
        self.assertEqual('-R "select[mem>1500] rusage[mem=1500,mytokens=42]" -M1500', resources)
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1.5, 'cmd', threads=42)

        bsub.memory_units='KB'
        resources = bsub._make_resources_string()
        self.assertEqual('-n 42 -R "span[hosts=1] select[mem>1500] rusage[mem=1500]" -M1500000', resources)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', tmp_space=42)
        bsub.memory_units='MB'
        resources = bsub._make_resources_string()
        self.assertEqual('-R "select[mem>1000 && tmp>42000] rusage[mem=1000,tmp=42000]" -M1000', resources)

    def test_make_queue_string(self):
        '''Check that queue set correctly'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        self.assertEqual('-q queue', bsub._make_queue_string())
        bsub = lsf.Job('out', 'error', 'name', None, 1, 'cmd')
        self.assertEqual('', bsub._make_queue_string())

    def test_make_output_files_string(self):
        '''Check that the names of the stdout and stderr files set properly'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        self.assertEqual('-o out -e error', bsub._make_output_files_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', array_start=1, array_end=42)
        self.assertEqual('-o out.%I -e error.%I', bsub._make_output_files_string())

    def test_make_job_name_string(self):
        '''Check that the name of the job is set correctly'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        self.assertEqual('-J name', bsub._make_job_name_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', array_start=1, array_end=42)
        self.assertEqual('-J "name[1-42]%100"', bsub._make_job_name_string())

    def test_make_dependencies_string(self):
        '''Check that -w 'done()' and 'ended()' are constructed correctly'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        self.assertEqual('', bsub._make_dependencies_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', depend=42)
        self.assertEqual("-w 'done(42)'", bsub._make_dependencies_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', depend=[42,'a'])
        self.assertEqual("-w 'done(42) && done(\"a\")'", bsub._make_dependencies_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', ended=42)
        self.assertEqual("-w 'ended(42)'", bsub._make_dependencies_string())

    def test_make_command_string(self):
        '''Check that command to be bsubbed is made correctly - including for job array'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        self.assertEqual('cmd', bsub._make_command_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd', checkpoint=True)
        self.assertEqual('cr_run cmd', bsub._make_command_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd.INDEX')
        self.assertEqual('cmd.INDEX', bsub._make_command_string())
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd.INDEX foo bar.INDEX', array_start=1, array_end=42)
        self.assertEqual('cmd.\$LSB_JOBINDEX foo bar.\$LSB_JOBINDEX', bsub._make_command_string())

    def test_add_dependency(self):
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub.add_dependency('jobname')
        dependencies = ['"jobname"']
        self.assertListEqual(bsub.run_when_done, dependencies)

        bsub.add_dependency(42)
        dependencies.append('42')
        self.assertListEqual(bsub.run_when_done, dependencies)

        bsub.add_dependency('43')
        dependencies.append('43')
        self.assertListEqual(bsub.run_when_done, dependencies)

        bsub.add_dependency(['jobname2', 44])
        dependencies.extend(['"jobname2"', '44'])
        self.assertListEqual(bsub.run_when_done, dependencies)

        self.assertListEqual(bsub.run_when_ended, [])

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub.add_dependency('jobname', ended=True)
        dependencies = ['"jobname"']
        self.assertListEqual(bsub.run_when_ended, dependencies)

        bsub.add_dependency(42, ended=True)
        dependencies.append('42')
        self.assertListEqual(bsub.run_when_ended, dependencies)

        bsub.add_dependency('43', ended=True)
        dependencies.append('43')
        self.assertListEqual(bsub.run_when_ended, dependencies)

        bsub.add_dependency(['jobname2', 44], ended=True)
        dependencies.extend(['"jobname2"', '44'])
        self.assertListEqual(bsub.run_when_ended, dependencies)

        self.assertListEqual(bsub.run_when_done, [])

    def test_set_job_id_from_bsub_output(self):
        '''Check that stdout from calling bsub is parsed correctly'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub_out = 'This is not good bsub output\nso should throw error\n'
        with self.assertRaises(lsf.Error):
            bsub._set_job_id_from_bsub_output(bsub_out)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        with self.assertRaises(lsf.Error):
            bsub._set_job_id_from_bsub_output('')

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        with self.assertRaises(lsf.Error):
            bsub._set_job_id_from_bsub_output(None)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub_out = 'Job <42> is submitted to queue <normal>.\nAnother line of output\n'
        bsub._set_job_id_from_bsub_output(bsub_out)
        self.assertEqual('42', bsub.job_id)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub_out = 'Job <not_an_int> is submitted to queue <normal>\n'
        with self.assertRaises(lsf.Error):
            bsub._set_job_id_from_bsub_output(bsub_out)

    def test_run(self):
        '''Test run() runs/fails as expected'''
        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._run_test_cmd = os.path.join(test_dir, 'lsf_unittest_run_bsub_fails.sh')
        with self.assertRaises(lsf.Error):
            bsub.run()

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._run_test_cmd = 'not_a_command_and_should_fail'
        with self.assertRaises(lsf.Error):
            bsub.run()

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'cmd')
        bsub._run_test_cmd = os.path.join(test_dir, 'lsf_unittest_run_bsub_ok.sh')
        bsub.run()
        self.assertEqual('42', bsub.job_id)

    def test_run_not_bsubbed(self):
        '''Test running not bsubbed on normal command'''
        tmp_out = 'tmp.test_run_not_bsubbed'
        bsub = lsf.Job('tmp.out', 'tmp.err', 'test', 'normal', 1, 'echo "test_run_not_bsubbed" > ' + tmp_out)
        bsub.run_not_bsubbed()
        self.assertTrue(filecmp.cmp(tmp_out, os.path.join(test_dir, 'lsf_unittest_run_not_bsubbed.out')))
        os.unlink(tmp_out)

        bsub = lsf.Job('out', 'error', 'name', 'queue', 1, 'not_a_command_and_should_fail')
        with self.assertRaises(lsf.Error):
            bsub.run_not_bsubbed()

    def test_run_not_bsubbed_array(self):
        '''Test running not bsubbed on job array'''
        tmp_out = 'tmp.test_run_not_bsubbed_array'
        bsub = lsf.Job('tmp.out', 'tmp.err', 'test', 'normal', 1, 'echo "test_run_not_bsubbed array INDEX" >> ' + tmp_out, array_start=1, array_end=3)
        bsub.run_not_bsubbed()
        self.assertTrue(filecmp.cmp(tmp_out, os.path.join(test_dir, 'lsf_unittest_run_not_bsubbed_array.out')))
        os.unlink(tmp_out)


if __name__ == '__main__':
    unittest.main()


farmpy/tasks.py
from farmpy import lsf_stats
import sys

class Error (Exception): pass


def lsf_out_to_tsv(infiles, outfile, show_all=False, compress_job_name=10, compress_filename=None, time_in_hours=False):
    '''Given a list of files out bsub output, makes a tsv file of their stats'''
    if outfile == '-':
        fout = sys.stdout
    else:
        try:
            fout = open(outfile, 'w')
        except:
            raise Error ('Error opening file "' + outfile + '"')

    if show_all:
        print('#number_in_file', lsf_stats.tsv_header, 'filename', sep='\t', file=fout)
    else:
        print('#number_in_file', lsf_stats.tsv_header_short, 'filename', sep='\t', file=fout)

    for infile in infiles:
        attempt_number = 1
        reader = lsf_stats.file_reader(infile)
        
        for stats in reader:
            if compress_filename is None:
                filename = infile
            elif len(infile) > compress_filename:
                filename = '*' + infile[-compress_filename:]
            else:
                filename = infile
            print(attempt_number, stats.to_tsv(job_name_limit=compress_job_name, show_all=show_all, time_in_hours=time_in_hours), filename, sep='\t', file=fout)
            attempt_number += 1

    if outfile != '-':
        fout.close()


farmpy/lsf.py
"""A module to help submitting jobs to a compute farm that uses Platform's LSF

Consists of one class - Job.
You can create jobs, run them and set jobs to depend on other jobs.

Run a job:
  job1 = Job('out', 'err', 'job name', 'normal', 1, 'run.sh')
  job1.run()

This creates a job called 'job name', which will run in the normal queue.
It asks for 1GB of memory, writes stdout/stderr to the files out/err.
It runs the script run.sh.

Dependencies:
You can use job1.add_dependency('another job name') to set job1 to run when the job with 'another job name' finishes. This means you can do this to set jobs to depend on other jobs:
  job2 = Job('out2', 'err2', 'job2', 'normal', 1, 'run2.sh')
  job2.add_dependency(job1.name)
  job2.run()
job2 will be put into the queue, but will only run when job1 finishes successfully.
Alternatively, running a job sets the variable job1.job_id, which means you can set a depndency is like this:
  job1.run() # this sets the variable job1.job_id.
  job2.add_dependency(job1.job_id)
This is safer than using a job name because the name might not be unique, but the ID is always unique.

You can also use Job(..., ended=job1.job_id) to make job2 run when job1 finishes, regardless of whether or not job1 finished successfully.

Job arrays:
You can run a job array using stert= and end= when constructing a Job. Example:
  job = ('out', 'err', 'name', 'normal', 1, 'run.sh INDEX', start=1, end=10)
This sets up a job array with 1 elements. stdout files will be called out.1, out.2, ....etc and similarly for stderr. Every appearance of 'INDEX' in the command is translated to be the job index (technically, the $LSB_JOBINDEX environment variable). So this would run submit 10 jobs to LSF:
  run.sh 1
  run.sh 2
  ...
  run.sh 10

Extra options:
When calling Job(...), these are options that can be given:
  start=x, end=y   - for running job arrays (se above)
  threads=N        - ask for N threads (default = 1)
  tmo_space=x      - ask for x GB of tmp space
  nax_array_size=N - limit number of jobs running at the same time in an array to N (default 100)
  memory_units=KB or MB - the units used in the -M option. It should be detected automatically, but you can override using this option (but might cause run() to fail)

A note on memory units:
The memory may need to be specified in KB or MB.
The units are determined by running `lsadmin showconf lim`.
However, in case that doesn't work (or you want to avoid running it), you
can set the environment variable FARMPY_LSF_MEMORY_UNITS to MB or KB.
"""

import socket
import subprocess
import getpass
import os


class Error (Exception): pass

class Job:
    def __init__(self, out, error, name, queue, mem, cmd,
                 array_start=0, array_end=0,
                 depend=None,
                 ended=None,
                 threads=1,
                 tmp_space=0,
                 memory_units=None,
                 no_resources=False,
                 checkpoint=False,
                 checkpoint_dir=None,
                 checkpoint_period=600,
                 tokens_name=None,
                 tokens_number=100,
                 max_array_size=100):
        '''Creates Job object. See main module help for a description and example usage'''

        self.stdout_file = out
        self.stderr_file = error
        self.name = name
        self.queue = queue
        self.memory = int(1000 * round(mem, 3))
        self.command = cmd
        self.array_start = array_start
        self.array_end = array_end
        self.run_when_done = []
        self.run_when_ended = []
        self.add_dependency(depend)
        self.add_dependency(ended, ended=True)
        self.threads = threads
        self.tmp_space = int(1000 * tmp_space)
        self.memory_units = memory_units
        self.max_array_size = max_array_size
        self.job_id = None
        self.no_resources=no_resources
        self.checkpoint = checkpoint
        self.checkpoint_dir = checkpoint_dir
        self.checkpoint_period = checkpoint_period
        self.tokens_name = tokens_name
        self.tokens_number = tokens_number


        # these are used for unittests to call test scripts instead of the
        # real commands you would run on a farm
        self._lsadmin_cmd = 'lsadmin showconf lim ' + socket.gethostname()
        self._run_test_cmd = None


    def run(self, verbose=False):
        '''Submits the job to the farm. Dies if not successful.'''
        if self._run_test_cmd is not None:
            cmd = self._run_test_cmd
        else:
            cmd = str(self)

        try:
            bsub_out = subprocess.check_output(cmd, shell=True).decode('utf-8')
        except:
            raise Error('Error in bsub call. I tried to run:\n' + str(self))

        self._set_job_id_from_bsub_output(bsub_out)


    def run_not_bsubbed(self):
        '''Runs the job directly on the node. Does not bsub it. Stdout and stderr will be output as if the command was run directly in a terminal'''
        def run_command(cmd):
            retcode = subprocess.call(cmd, shell=True)
            if retcode != 0:
                raise Error('Error running command:\n' + str(self))

        if self.array_start > 0:
            for i in range(self.array_start, self.array_end + 1):
                os.environ['LSB_JOBINDEX'] = str(i)
                run_command(self._make_command_string().replace('\$LSB_JOBINDEX', '$LSB_JOBINDEX'))
        else:
            run_command(self._make_command_string())


    def add_dependency(self, deps, ended=False):
        '''Makes the job depend on another job or jobs.

        deps -- can either be a job name, job id, or a list of job names and/or ids. It is assumed that anything that is all digits is a job id, otherwise it is a job name.

        Default is to make the job only run when the jobs it depends on finish successfully (i.e. return zero error code). If you want this job to run when the job it depends on finishes, regardless or error code, set ended=True.'''
        if deps is None:
            return

        if type(deps) is not list:
            deps = [deps]

        for d in deps:
            try:
                x = int(d)
                x = str(d)
            except ValueError:
                x = '"' + d + '"'

            if ended:
                self.run_when_ended.append(x)
            else:
                self.run_when_done.append(x)

    def _make_checkpoint_string(self):
        if self.checkpoint:
            if self.checkpoint_dir is None:
                self.checkpoint_dir = self.stdout_file + '.checkpoint'

            return '-k "' + os.path.abspath(self.checkpoint_dir) + ' method=blcr ' + str(self.checkpoint_period) + '"'
        else:
            return ''

    def _make_prexec_test_string(self):
        # a reasonable sanity check is that the home directory of the user
        # running the job is mounted on the node that will run the job
        return "-E 'test -e " + os.path.expanduser('~') + "'"


    def _make_queue_string(self):
        return '' if self.queue is None else '-q ' + self.queue


    def _set_memory_units(self):
        if self.memory_units is not None:
            return self.memory_units
        elif 'FARMPY_LSF_MEMORY_UNITS' in os.environ:
            self.memory_units = os.environ['FARMPY_LSF_MEMORY_UNITS']
        else:
            try:
                output = subprocess.check_output(self._lsadmin_cmd, shell=True).decode('utf-8').rstrip().split('\n')
            except:
                raise Error("Error getting LSF memory units using: lsadmin showconf lim " + socket.gethostname() + '\n... you can work around this by setting the environment variable FARMPY_LSF_MEMORY_UNITS to KB or MB, as appropriate for your LSF setup.')

            # get the line with LSF_UNIT_FOR_LIMITS in it, if it exists
            for line in output:
                data = line.strip().split()
                if data[0] == 'LSF_UNIT_FOR_LIMITS':
                    self.memory_units = data[2]
                    break
            else:
                self.memory_units = 'KB'

        if self.memory_units not in ['KB', 'MB']:
            raise Error('Error getting lsf memory units. Expected KB or MB')


    def _make_resources_string(self):
        if self.no_resources:
            return ''

        self._set_memory_units()
        s = ''

        if self.threads > 1:
            s += '-n ' + str(self.threads) + ' -R "span[hosts=1]' + ' '
        else:
            s = '-R "'

        s += 'select[mem>' + str(self.memory)

        if self.tmp_space:
            s += ' && tmp>' + str(self.tmp_space)

        s += '] rusage[mem=' + str(self.memory)

        if self.tmp_space:
            s += ',tmp=' + str(self.tmp_space)

        if self.tokens_name:
            s += ',' + self.tokens_name + '=' + str(self.tokens_number)

        s += ']" -M' + str(self.memory)

        if self.memory_units == 'KB':
            s += '000'

        return s


    def _make_output_files_string(self):
        if self.array_start > 0:
            return '-o ' + self.stdout_file + '.%I -e ' + self.stderr_file + '.%I'
        else:
            return '-o ' + self.stdout_file + ' -e ' + self.stderr_file


    def _make_job_name_string(self):
        if self.array_start > 0:
            return '-J "' + self.name + '[' + str(self.array_start) + '-' + str(self.array_end) \
                   + ']%' + str(self.max_array_size) + '"'
        else:
            return '-J ' + self.name


    def _make_dependencies_string(self):
        if len(self.run_when_done) + len(self.run_when_ended) > 0:
            return "-w '" \
                   + ' && '.join(['done(' + x + ')' for x in self.run_when_done] + ['ended(' + x + ')' for x in self.run_when_ended]) \
                   + "'"
        else:
            return ''


    def _make_command_string(self):
        command = ''

        if self.checkpoint:
            command = 'cr_run '

        if self.array_start > 0:
            return command + self.command.replace('INDEX', '\$LSB_JOBINDEX')
        else:
            return command + self.command


    def __str__(self):
        return ' '.join([x for x in [
                            'bsub',
                            self._make_checkpoint_string(),
                            self._make_queue_string(),
                            self._make_prexec_test_string(),
                            self._make_resources_string(),
                            self._make_output_files_string(),
                            self._make_job_name_string(),
                            self._make_dependencies_string(),
                            self._make_command_string()
                        ] if x != ''])


    def _set_job_id_from_bsub_output(self, bsub_output):
        # for ok bsub call, output should look like this:
        # Job <42> is submitted to queue <normal>.
        if bsub_output is None:
            raise Error('Error - got no output from bsub call')

        first_line = bsub_output.split('\n')[0]

        if 'is submitted to' not in first_line:
            raise Error("Error getting job ID from bsub output. I got this:\n" + bsub_output)

        self.job_id = first_line.split()[1][1:-1]

        # check that the job ID looks like an int
        try:
            int(self.job_id)
        except ValueError:
            raise Error('Error getting job ID from this:' + first_line)

